### PR TITLE
generalize lookup to also handle vector

### PR DESCRIPTION
**Pull request by MichaelPFrey**
as discussed in #4270.

One use case is for interpolating between keypoints of an animation.

An other use case is something like this:

```
n = 7;

function pointOnCircle(r,a,x,y) = [r * cos(a) + x, r * sin(a) + y];

a = [ for(i=[0:1:n]) [i/n, pointOnCircle(500,i*360/n*5,0,0)]];

echo(a);

 function get_cylinder_h(p) = lookup(p, a);

 color("black")
 for (i = [0:0.005:n]) {
    // echo(i, get_cylinder_h(i));
    echo(i);
    translate(get_cylinder_h(i)) cylinder(r1 = 6, r2 = 2, h = 30);
 }
```

Screenshot:

Screenshot from 2022-05-29 22-29-34

### CODE REVIEW

File: tests/data/scad/misc/lookup-tests.scad [hunk 1]
```diff
@@ -15,3 +15,23 @@ indices = [-20,-10,-9.9, -0.5, 0, 0.3, 1.1, 10, 10.1];
...
+
+table3DVector =
+        [[-1,  [-5, 5, -1]],
+         [-10, [-55, 55, -10]],
+         [0,    [0, 0, 0]],
+         [1,    [3, -3, 1]],
+         [10,   [333, -333, 10]]];
+for (i=[0:len(indices)-1]) {
+  echo(lookup(indices[i], table3DVector));
+}
```
Comment (jordanbrown0) [inline tests/data/scad/misc/lookup-tests.scad:37]: Add tests for error cases:

- [ ] key is not a number
- [ ] table is not a vector
- [ ] entry in table is not a vector
- [ ] entry in table has size other than 2
- [ ] entry[0] is not a number
- For the two entries bracketing a selected entry:
- [ ] entry[1] is neither a number nor a vector
- [ ] entry[1] is a vector containing values that are neither numbers nor vectors.
- [ ] the two entry[1] values are not consistent types - one is a number and the other is a vector.
- [ ] the two entry[1] values are both vectors, but they have different lengths.
- [ ] the two entry[1] values are both vectors, but one has an entry that is a number while the corresponding entry in the other is a vector.
- For the high or low entry, selected when the entry is out of bounds:
- [ ] [1] is not a number or vector.
- [ ] [1] is a vector, but it contains something other than a number or vector.  (Note:  recursively checking this is probably out of scope.)

Note:  a not-particularly-silly way to handle the type mismatch cases on the boundary cases might be to multiply them by 1.0.  That will do nothing if they are OK, and otherwise will trigger the same error handling that the interpolation cases would trigger, so you would at least sort of consistently get undef or warnings.

I suspect that most of those cases will yield undef and that a few will yield warnings.  I'm not looking for any behavior in particular, just that they not yield apparently-valid values and that they not yield crashes.

**Review by jordanbrown0 — CHANGES_REQUESTED** [review]

File: src/core/builtin_functions.cc [hunk 2]
```diff
@@ -441,31 +441,42 @@ Value builtin_lookup(Arguments arguments, const Location& loc)
...
       if (this_p >= p && (this_p < high_p || high_p < p)) {
         high_p = this_p;
-        high_v = this_v;
+        high_v = it;
       }
     }
+    if (p <= low_p) return high_v->toVector()[1].clone();
+    if (p >= high_p) return low_v->toVector()[1].clone();
+    double f = (p - low_p) / (high_p - low_p);
+    return (high_v->toVector()[1].operator*(f) + low_v->toVector()[1].operator*(1 - f));
```
[the remark this answers is not in the document]
Comment (thehans) [reply src/core/builtin_functions.cc:476]: The first is more mathematically robust.  Namely, when f == 1.0, it will return exactly the "high" value.  
This is a very common "lerp" operation, and you can find lots of discussions and Q&As online where the first expression is preferred.

[the remark this answers is not in the document]
Comment (jordanbrown0) [reply src/core/builtin_functions.cc:476]: OK

**Comment by rcolyer** [comment]
I ran a bunch of tests for robustness and function, and I think it is reasonably solid and a pretty useful addition.  translate(lookup($t, table)) for example is an animation game changer!  However, there are a few edge cases to address regarding bad inputs.  Following onto jordanbrown0's suggested test-cases, I think it might be pretty important to provide a warning whenever it attempts to interpolate between incompatible types.  Interpolation is only defined for number, or for vectors with numbers in them, and not for combinations of these.  The routine currently only type-checks the first entry in the look-up table, so all sorts of shenanigans can be passed in there following it.  So if it is not an exact entry from the table, I think at least a warning should result if a valid result cannot be obtained by the operation between the two elements.  This aligns with the philosophy of the additional warnings that have been added for undefined operations elsewhere, but none are produced in this PR at present.  Here it might be helpful to provide one and be a little explicit with the text in it about what went wrong, because with the added complexity in these tables, it could otherwise be a tricky problem to understand and spot.

**Comment by jordanbrown0** [comment]
This is on my list when I accumulate enough round tuits.
